File: discovery/models/discovery.py
"""Main discovery result models"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum

from .domain import DomainInfo
from .service import Service, Technology
from .finding import Finding


class DiscoveryStage(str, Enum):
    """Discovery pipeline stages"""
    INITIALIZED = "initialized"
    PASSIVE_DISCOVERY = "passive_discovery"
    ACTIVE_DISCOVERY = "active_discovery"
    DEEP_DISCOVERY = "deep_discovery"
    ENRICHMENT = "enrichment"
    COMPLETED = "completed"
    FAILED = "failed"


class TimelineEvent(BaseModel):
    """Timeline event during discovery"""
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    stage: DiscoveryStage
    event: str = Field(..., description="Event description")
    details: Optional[Dict[str, Any]] = Field(default=None, description="Additional details")


class DiscoveryMetadata(BaseModel):
    """Metadata about the discovery execution"""
    target: str = Field(..., description="Target URL/domain")
    scan_id: str = Field(..., description="Unique scan identifier")
    start_time: datetime = Field(default_factory=datetime.utcnow)
    end_time: Optional[datetime] = None
    duration_seconds: Optional[float] = None
    discovery_depth: str = Field(default="normal", description="Discovery depth level")
    tool_versions: Dict[str, str] = Field(
        default_factory=dict,
        description="Versions of tools used"
    )
    status: DiscoveryStage = Field(default=DiscoveryStage.INITIALIZED)
    error: Optional[str] = Field(default=None, description="Error message if failed")


class Endpoint(BaseModel):
    """Discovered API endpoint or URL path"""
    url: str = Field(..., description="Full endpoint URL")
    method: str = Field(default="GET", description="HTTP method")
    status_code: Optional[int] = Field(default=None, description="Response status")
    discovered_via: str = Field(..., description="Discovery source: crawl|js_analysis|sitemap|etc")
    parameters: List[str] = Field(default_factory=list, description="URL parameters")
    requires_auth: Optional[bool] = Field(default=None, description="Requires authentication")


class Recommendation(BaseModel):
    """Pentest focus recommendation"""
    category: str = Field(..., description="Recommendation category")
    priority: str = Field(..., description="Priority: high|medium|low")
    areas: List[str] = Field(..., description="Specific areas to focus on")
    rationale: Optional[str] = Field(default=None, description="Why this is recommended")


class Statistics(BaseModel):
    """Discovery statistics summary"""
    total_subdomains: int = 0
    live_services: int = 0
    technologies_detected: int = 0
    endpoints_discovered: int = 0
    findings_by_severity: Dict[str, int] = Field(
        default_factory=lambda: {
            "critical": 0,
            "high": 0,
            "medium": 0,
            "low": 0,
            "info": 0
        }
    )
    coverage_completeness: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Estimated discovery completeness"
    )


class DiscoveryResult(BaseModel):
    """Complete discovery result output"""
    metadata: DiscoveryMetadata = Field(..., description="Scan metadata")
    discovery_timeline: List[TimelineEvent] = Field(
        default_factory=list,
        description="Timeline of discovery events"
    )
    domains: Optional[DomainInfo] = Field(default=None, description="Domain intelligence")
    services: List[Service] = Field(default_factory=list, description="Live web services")
    technologies: List[Technology] = Field(default_factory=list, description="All detected technologies")
    endpoints: List[Endpoint] = Field(default_factory=list, description="Discovered endpoints")
    findings: List[Finding] = Field(default_factory=list, description="Security findings")
    recommendations: List[Recommendation] = Field(
        default_factory=list,
        description="Pentest focus recommendations"
    )
    statistics: Statistics = Field(default_factory=Statistics, description="Summary statistics")

    def add_timeline_event(self, stage: DiscoveryStage, event: str, details: Optional[Dict[str, Any]] = None):
        """Add event to timeline"""
        self.discovery_timeline.append(
            TimelineEvent(stage=stage, event=event, details=details)
        )

    def update_statistics(self):
        """Recalculate statistics from current data"""
        if self.domains:
            self.statistics.total_subdomains = len(self.domains.subdomains)

        self.statistics.live_services = len(self.services)
        self.statistics.technologies_detected = len(self.technologies)
        self.statistics.endpoints_discovered = len(self.endpoints)

        # Count findings by severity
        severity_counts = {"critical": 0, "high": 0, "medium": 0, "low": 0, "info": 0}
        for finding in self.findings:
            severity_counts[finding.severity.value] += 1
        self.statistics.findings_by_severity = severity_counts

        # Estimate completeness (simple heuristic for PoC)
        completeness_score = 0.0
        if self.domains and self.domains.total_subdomains > 0:
            completeness_score += 0.3
        if self.statistics.live_services > 0:
            completeness_score += 0.3
        if self.statistics.endpoints_discovered > 10:
            completeness_score += 0.2
        if self.statistics.technologies_detected > 0:
            completeness_score += 0.2

        self.statistics.coverage_completeness = min(completeness_score, 1.0)

File: src/models/discovery.ts
/**
 * Main discovery result models
 */

import { z } from 'zod';
import { domainInfoSchema } from './domain.js';
import { serviceSchema, technologySchema } from './service.js';

/**
 * Discovery pipeline stages
 */
export enum DiscoveryStage {
  INITIALIZED = 'initialized',
  PASSIVE_DISCOVERY = 'passive_discovery',
  PORT_DISCOVERY = 'port_discovery',
  ACTIVE_DISCOVERY = 'active_discovery',
  DEEP_DISCOVERY = 'deep_discovery',
  ENRICHMENT = 'enrichment',
  AUTHENTICATED_DISCOVERY = 'authenticated_discovery',
  COMPLETED = 'completed',
  FAILED = 'failed',
}

export const discoveryStageSchema = z.nativeEnum(DiscoveryStage);

/**
 * Timeline event during discovery
 */
export const timelineEventSchema = z.object({
  /** Event timestamp */
  timestamp: z.date().default(() => new Date()),
  /** Discovery stage */
  stage: discoveryStageSchema,
  /** Event description */
  event: z.string(),
  /** Event message (alias for event) */
  message: z.string().optional(),
  /** Additional details */
  details: z.record(z.any()).optional(),
});

export type TimelineEvent = z.infer<typeof timelineEventSchema>;

/**
 * Metadata about the discovery execution
 */
export const discoveryMetadataSchema = z.object({
  /** Target URL/domain */
  target: z.string(),
  /** Unique scan identifier */
  scanId: z.string(),
  /** Scan start time */
  startTime: z.date().default(() => new Date()),
  /** Scan end time */
  endTime: z.date().optional(),
  /** Duration in seconds */
  durationSeconds: z.number().nonnegative().optional(),
  /** Discovery depth level */
  discoveryDepth: z.enum(['shallow', 'normal', 'deep']).default('normal'),
  /** Versions of tools used */
  toolVersions: z.record(z.string()).default({}),
  /** Current discovery status */
  status: discoveryStageSchema.default(DiscoveryStage.INITIALIZED),
  /** Error message if failed */
  error: z.string().optional(),
});

export type DiscoveryMetadata = z.infer<typeof discoveryMetadataSchema>;

/**
 * Discovered API endpoint or URL path
 */
export const endpointSchema = z.object({
  /** Full endpoint URL */
  url: z.string().url(),
  /** HTTP method */
  method: z.string().default('GET'),
  /** Response status code */
  statusCode: z.number().int().min(100).max(599).optional(),
  /** Discovery source: crawl|js_analysis|sitemap|etc */
  discoveredVia: z.string(),
  /** URL parameters */
  parameters: z.array(z.string()).default([]),
  /** Requires authentication */
  requiresAuth: z.boolean().optional(),
});

export type Endpoint = z.infer<typeof endpointSchema>;

/**
 * Pentest focus recommendation
 */
export const recommendationSchema = z.object({
  /** Recommendation category */
  category: z.string(),
  /** Priority: high|medium|low */
  priority: z.enum(['high', 'medium', 'low']),
  /** Specific areas to focus on */
  areas: z.array(z.string()),
  /** Why this is recommended */
  rationale: z.string().optional(),
});

export type Recommendation = z.infer<typeof recommendationSchema>;

/**
 * Discovery statistics summary
 */
export const statisticsSchema = z.object({
  /** Total subdomains discovered */
  totalSubdomains: z.number().int().nonnegative().default(0),
  /** Live/responsive services */
  liveServices: z.number().int().nonnegative().default(0),
  /** Unique technologies detected */
  technologiesDetected: z.number().int().nonnegative().default(0),
  /** Total endpoints discovered */
  endpointsDiscovered: z.number().int().nonnegative().default(0),
  /** Total endpoints (alias for compatibility) */
  totalEndpoints: z.number().int().nonnegative().default(0),
  /** Total forms discovered */
  totalForms: z.number().int().nonnegative().default(0),
  /** Authenticated endpoints count */
  authenticatedEndpoints: z.number().int().nonnegative().default(0),
  /** Total ports scanned across all hosts */
  totalPortsScanned: z.number().int().nonnegative().default(0),
  /** Total open ports discovered */
  openPortsFound: z.number().int().nonnegative().default(0),
  /** Total open ports (alias for compatibility) */
  totalOpenPorts: z.number().int().nonnegative().default(0),
  /** Number of hosts with open ports */
  hostsWithOpenPorts: z.number().int().nonnegative().default(0),
  /** Estimated discovery completeness (0.0 - 1.0) */
  coverageCompleteness: z.number().min(0.0).max(1.0).default(0.0),
});

export type Statistics = z.infer<typeof statisticsSchema>;

/**
 * Complete discovery result output
 */
export const discoveryResultSchema = z.object({
  /** Scan metadata */
  metadata: discoveryMetadataSchema,
  /** Timeline of discovery events */
  discoveryTimeline: z.array(timelineEventSchema).default([]),
  /** Timeline (alias for discoveryTimeline) */
  timeline: z.array(timelineEventSchema).optional(),
  /** Domain intelligence */
  domains: domainInfoSchema.optional(),
  /** Live web services */
  services: z.array(serviceSchema).default([]),
  /** All detected technologies */
  technologies: z.array(technologySchema).default([]),
  /** Discovered endpoints */
  endpoints: z.array(endpointSchema).default([]),
  /** Pentest focus recommendations */
  recommendations: z.array(recommendationSchema).default([]),
  /** Infrastructure intelligence (cloud providers, CDN, ASN) */
  infrastructure: z.record(z.any()).default({}),
  /** Summary statistics */
  statistics: statisticsSchema.default({}),
});

export type DiscoveryResult = z.infer<typeof discoveryResultSchema>;

/**
 * Helper functions for discovery results
 */
export class DiscoveryResultHelper {
  /**
   * Add event to timeline
   */
  static addTimelineEvent(
    result: DiscoveryResult,
    stage: DiscoveryStage,
    event: string,
    details?: Record<string, unknown>
  ): DiscoveryResult {
    return {
      ...result,
      discoveryTimeline: [
        ...result.discoveryTimeline,
        {
          timestamp: new Date(),
          stage,
          event,
          details,
        },
      ],
    };
  }

  /**
   * Recalculate statistics from current data
   */
  static updateStatistics(result: DiscoveryResult): DiscoveryResult {
    const statistics: Statistics = { ...result.statistics };

    if (result.domains) {
      statistics.totalSubdomains = result.domains.subdomains.length;

      // Calculate port statistics
      let totalPortsScanned = 0;
      let totalOpenPorts = 0;
      let hostsWithOpenPorts = 0;

      for (const subdomain of result.domains.subdomains) {
        totalPortsScanned += subdomain.totalPortsScanned;
        totalOpenPorts += subdomain.openPortsCount;
        if (subdomain.openPortsCount > 0) {
          hostsWithOpenPorts++;
        }
      }

      statistics.totalPortsScanned = totalPortsScanned;
      statistics.openPortsFound = totalOpenPorts;
      statistics.hostsWithOpenPorts = hostsWithOpenPorts;
    }

    statistics.liveServices = result.services.length;
    statistics.technologiesDetected = result.technologies.length;
    statistics.endpointsDiscovered = result.endpoints.length;

    // Estimate completeness (simple heuristic)
    let completenessScore = 0.0;
    if (result.domains && result.domains.totalSubdomains > 0) {
      completenessScore += 0.3;
    }
    if (statistics.liveServices > 0) {
      completenessScore += 0.3;
    }
    if (statistics.endpointsDiscovered > 10) {
      completenessScore += 0.2;
    }
    if (statistics.technologiesDetected > 0) {
      completenessScore += 0.2;
    }

    statistics.coverageCompleteness = Math.min(completenessScore, 1.0);

    return {
      ...result,
      statistics,
    };
  }
}
